### PR TITLE
Swedev 414443

### PR DESCRIPTION
**Pull request by petar-avramovic**

### CODE REVIEW

File: llvm/lib/Target/AMDGPU/SIInstrInfo.cpp [hunk 1]
```diff
@@ -171,6 +171,39 @@ bool SIInstrInfo::isIgnorableUse(const MachineOperand &MO) const {
          isVALU(*MO.getParent()) && !resultDependsOnExec(*MO.getParent());
 }
 
+bool SIInstrInfo::modifiesRegisterImplicitly(
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/SIInstrInfo.cpp:174]: We have an ever growing set of convoluted callbacks to deal with this problem. I think we need to look into moving some of the modeling of this into the CFG itself. I think we need to mark block edges as convergent/divergent/something

File: llvm/lib/Target/AMDGPU/SIInstrInfo.cpp [hunk 2]
```diff
@@ -171,6 +171,39 @@ bool SIInstrInfo::isIgnorableUse(const MachineOperand &MO) const {
...
+    // bb.B
+    //  SI_END_CF %EndCF, implicit-def dead $exec
+    //  ... MoveCandidate should be moved here
+
+    // MoveCandidate is from block that started divergent control flow via SI_IF
+    // it is a simple SI_IF (no loops) - only user of SI_IF is SI_END_CF
+    // SI_END_CF restores exec mask as it was before SI_IF (unchanged)
+    Register EndCF = ModifierInstr->getOperand(0).getReg();
+    MachineInstr *SIIF = MRI.getVRegDef(EndCF);
+    if (SIIF->getOpcode() == AMDGPU::SI_IF && MRI.hasOneUse(EndCF) &&
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/SIInstrInfo.cpp:199]: This doesn't even work after lowering

**Comment by arsenm** [comment]
Description needs to be what this is doing, the swdev thing should be a comment footer

**Comment by arsenm** [comment]
Is this still relevant? I thought a newer patch fixed the same issue?

**Comment by petar-avramovic** [comment]
Implemented in https://github.com/llvm/llvm-project/pull/67456

**Pull request closed without merging**